README
X Letters 2 (Unicode) - Catch falling words

This version 2 is a major rewrite, so expect bugs and broken features
(with respect to xletters versions 1.1.x). Some features described in
the doc (e.g. the possibility to select dictionary entries according
to difficulty levels) are not implemented yet. 
Below you can find: 
* install instructions
* comments about new features and un-features

----------------------------------------------------------
* warnings for an installation alongside an old xletters (version <= 1.1.x)

Beware that any previous version of xletters will share its application
defaults with this one.

----------------------------------------------------------
* install instructions

0. Read the GPL carefully if you don't know it by heart already.  You
can find it in the file COPYING.  It applies to xletters.

1. Edit the file xletters.h; you can leave everything untouched if you
want.

2. Build xletters by typing
	./configure && make

3. Become root and install xletters by typing
	make install
You may want to make xletters sgid games. Read the INSTALL
file for further information.
Except for the score file, everything will go into /usr/local by
default. A directory /var/local/games/lib/xletters is created, with a 
world-readable empty score file inside.

				Paris, 2004/01/06
				Adrian Daerr

----------------------------------------------------------
* comments about new features and un-features

(see also Changelog)

- Any UTF-8 encoded word list or dictionary may be used. See the file
"dictionary.html" for information about the file format. Word-lists
for xletters v<=1.1.x will continue to work. See the included
Japanese-english dictionary "jp-en.txt" as an example of xletters new
capabilities (distinction between drawn and typed characters, input
conversion layer). If you create or find a word list or dictionary you
would like to share, please drop a note.

The included "jp-en.txt" is a processed version of the
Japanese-English Electronic Dictionary File EDICT 25APR02 V02-001, and
is copyright J.W. Breen - 2002).

- The tranlation tables to convert keyboard input to typed characters
(in the case of japanese: kana) are contained in the file
"inputconv.txt". You can edit this file with any editor capable of
reading and saving Unicode in UTF-8 format.

- New options:
  -showk n, which says if and how the kana transcription should be shown
    next to the kanji. The value of the integer parameter n is
    somewhere between 0 (don't show furigana transcription) and 31
    (preselect transcription and show everything you can) and is
    obtained by summing the numbers in front of the desired behaviour:
     1 show typed part of furigana, in the lower part of the screen
     2 show typed part of furigana, always
     4 show untyped part of furigana, in the lower part of the screen
     8 show untyped part of furigana, always
    16 pick one furigana transcription at the beginning (i.e. 
       to destroy word, only this transcription will work). This implies +8 
       to avoid frustration.

  -showc n, which says if and how the comment/translation should be shown:
    n=0 don't show comment/translation
    n=1 show comment/translation, in the lower part of the screen
    n=2 show comment/translation, always

- Don't try to put CONTEXT = 1 in xletters.h. It won't work.

- The deathmatch and duel modes are broken. Sorry.


Changelog
2004-1-11  Extended inputconv.txt to be much more liberal (and corrected a 
 2.0.4     bug for katakana-dzu). Now hopefully accepts everything existing
           input-methods accept (and more), e.g.  double consonnants for
           small-tsu preceding a consonnant, etc. 
           Added (still to inputconv.txt) a compositing table for western
           european languages (incomplete).

2004-1-6   Included jp-en.txt pre-processed to include markers and 
 2.0.3     meta-information (see CompJapanEngDict.java).
           Attempt to merge with Felipe Bergo (bergo@seul.org)'s 1.1.1
           branch: *adding automake/autoconf building,
             (binary will be called "xletters2" until code becomes 
             reasonably stable)
                   *renaming config.h to xletters.h

2003-7-21  Fixed big SEGFAULT bug in 2.1.  When running into 
 2.0.2     difficulties at establishing the kana2kanji correspondence,
 (alias    xletters now prints the word-file line number where player need 
 jp-2.2)   to help with markers.

2003-7-21  major rewrite: handles kanji words; words may have several
 2.0.1     kana-transcriptions, plus possibly a translation/comment;
 (alias    dictionary file is now UTF-8 Unicode; word lists for original
 jp-2.1)   xletters (v<=1.10) are accepted; input-conversion rules are 
           read at runtime, also UTF-8 format; ...

2003-5-17  (version naming changed to avoid confusion with new 
 2.0.0     potential normal xletters versions)
 (alias    merged romajikana.cnv and ascii_jis0208.txt
 jp-2.0)   minor changes to ascii_jis0208 translation table

2003-5-12  first xletters version patched for Japanese dictionnary
 1.1.0jp   starting point: original xletters version 1.1.0


utf8.h
/* The original file utf8.h was taken from the source of the
 * GNU LIBICONV Library and slightly modified. */

/*
 * Copyright (C) 1999-2001 Free Software Foundation, Inc. 
 * Copyright (C) 2003 Adrian Daerr 
 * This file is part of xletters. It is distributed under the terms of
 * the GNU Public License, a copy of which should be with this
 * software.
 *
 * xletters is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * xletters is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xletters; see the file COPYING.txt. If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/*
 * UNICODE
 */

typedef u_int32_t ucs4_t;
typedef ucs4_t wchar;
#define UCS4_NULL (ucs4_t)0
#define UCS4_LF (ucs4_t)0xa
#define UCS4_SPC (ucs4_t)0x20
/* #define UCS4_POINT (ucs4_t)0x2e */
#define UCS4_POINT (ucs4_t)0x2d /* make hyphen */
#define UCS4_SLASH (ucs4_t)0x2f 
#define UCS4_LT (ucs4_t)0x3c
#define UCS4_GT (ucs4_t)0x3e 
#define UCS4_LSQUAREBRACKET (ucs4_t)0x5b
#define UCS4_RSQUAREBRACKET (ucs4_t)0x5d
/* following is no good if machine uses e.g. EBDIC instead of ASCII */
#define UCS4(x) ((ucs4_t)(unsigned char)x)
#define ASCII(x) ((char)(x & 0xff))

/*
 * UTF-8
 */

#define RET_TOOFEW -1 /* to few chars in input stream */
#define RET_TOOSMALL -2 /* not enough space in output buffer */
#define RET_ILSEQ -3 /* not a permissible UTF-8 sequence */
#define RET_ILUNI -4 /* not a permissible UCS-4 character */

/* Specification: RFC 2279 */

/* The following function converts a UTF-8 coded Unicode character
   (pointed to by s) into a UCS-4 coded character (pointed to by pwc).
   The argument n indicates how many chars are available in the input
   stream s. The return value indicates exactly how many chars were
   consumed from the UTF-8 stream at s. If the end of stream is
   reached before the UCS-4 character is complete, the function
   returns RET_TOOFEW. If the stream does not comply with RFC 2279,
   RET_ILSEQ is returned. */
static int utf8_mbtowc (ucs4_t *pwc, const unsigned char *s, int n)
{
  unsigned char c = s[0];

  if (c < 0x80) {
    *pwc = c;
    return 1;
  } else if (c < 0xc2) {
    return RET_ILSEQ;
  } else if (c < 0xe0) {
    if (n < 2)
      return RET_TOOFEW;
    if (!((s[1] ^ 0x80) < 0x40))
      return RET_ILSEQ;
    *pwc = ((ucs4_t) (c & 0x1f) << 6)
           | (ucs4_t) (s[1] ^ 0x80);
    return 2;
  } else if (c < 0xf0) {
    if (n < 3)
      return RET_TOOFEW;
    if (!((s[1] ^ 0x80) < 0x40 && (s[2] ^ 0x80) < 0x40
          && (c >= 0xe1 || s[1] >= 0xa0)))
      return RET_ILSEQ;
    *pwc = ((ucs4_t) (c & 0x0f) << 12)
           | ((ucs4_t) (s[1] ^ 0x80) << 6)
           | (ucs4_t) (s[2] ^ 0x80);
    return 3;
  } else if (c < 0xf8 && sizeof(ucs4_t)*8 >= 32) {
    if (n < 4)
      return RET_TOOFEW;
    if (!((s[1] ^ 0x80) < 0x40 && (s[2] ^ 0x80) < 0x40
          && (s[3] ^ 0x80) < 0x40
          && (c >= 0xf1 || s[1] >= 0x90)))
      return RET_ILSEQ;
    *pwc = ((ucs4_t) (c & 0x07) << 18)
           | ((ucs4_t) (s[1] ^ 0x80) << 12)
           | ((ucs4_t) (s[2] ^ 0x80) << 6)
           | (ucs4_t) (s[3] ^ 0x80);
    return 4;
  } else if (c < 0xfc && sizeof(ucs4_t)*8 >= 32) {
    if (n < 5)
      return RET_TOOFEW;
    if (!((s[1] ^ 0x80) < 0x40 && (s[2] ^ 0x80) < 0x40
          && (s[3] ^ 0x80) < 0x40 && (s[4] ^ 0x80) < 0x40
          && (c >= 0xf9 || s[1] >= 0x88)))
      return RET_ILSEQ;
    *pwc = ((ucs4_t) (c & 0x03) << 24)
           | ((ucs4_t) (s[1] ^ 0x80) << 18)
           | ((ucs4_t) (s[2] ^ 0x80) << 12)
           | ((ucs4_t) (s[3] ^ 0x80) << 6)
           | (ucs4_t) (s[4] ^ 0x80);
    return 5;
  } else if (c < 0xfe && sizeof(ucs4_t)*8 >= 32) {
    if (n < 6)
      return RET_TOOFEW;
    if (!((s[1] ^ 0x80) < 0x40 && (s[2] ^ 0x80) < 0x40
          && (s[3] ^ 0x80) < 0x40 && (s[4] ^ 0x80) < 0x40
          && (s[5] ^ 0x80) < 0x40
          && (c >= 0xfd || s[1] >= 0x84)))
      return RET_ILSEQ;
    *pwc = ((ucs4_t) (c & 0x01) << 30)
           | ((ucs4_t) (s[1] ^ 0x80) << 24)
           | ((ucs4_t) (s[2] ^ 0x80) << 18)
           | ((ucs4_t) (s[3] ^ 0x80) << 12)
           | ((ucs4_t) (s[4] ^ 0x80) << 6)
           | (ucs4_t) (s[5] ^ 0x80);
    return 6;
  } else
    return RET_ILSEQ;
}

/* The following function converts the UCS-4 Unicode character wc into
   a UTF-8 sequence (stored in memory beginning at r). The argument n
   indicates how many chars may be output at most (n == 0 is
   acceptable). The return value indicates exactly how many chars were
   produced at r. There is not enough room in the output buffer at r
   (i.e. wc needs more than n chars in UTF-8), the function returns
   RET_TOOSMALL. If wc is not a valid Unicode character, the function
   returns RET_ILUNI. */
static int utf8_wctomb (unsigned char *r, ucs4_t wc, int n)
{
  int count;
  if (wc < 0x80)
    count = 1;
  else if (wc < 0x800)
    count = 2;
  else if (wc < 0x10000)
    count = 3;
  else if (wc < 0x200000)
    count = 4;
  else if (wc < 0x4000000)
    count = 5;
  else if (wc <= 0x7fffffff)
    count = 6;
  else
    return RET_ILUNI;
  if (n < count)
    return RET_TOOSMALL;
  switch (count) { /* note: code falls through cases! */
    case 6: r[5] = 0x80 | (wc & 0x3f); wc = wc >> 6; wc |= 0x4000000;
    case 5: r[4] = 0x80 | (wc & 0x3f); wc = wc >> 6; wc |= 0x200000;
    case 4: r[3] = 0x80 | (wc & 0x3f); wc = wc >> 6; wc |= 0x10000;
    case 3: r[2] = 0x80 | (wc & 0x3f); wc = wc >> 6; wc |= 0x800;
    case 2: r[1] = 0x80 | (wc & 0x3f); wc = wc >> 6; wc |= 0xc0;
    case 1: r[0] = wc;
  }
  return count;
}


CompJapanEngDict.java
/** Analyze Japanese words and establish a correspondence between the
 * Kanji writing and the Kana reading, based on the readings of the
 * Kanji making up the word. To run this on your own Japanese-XX
 * dictionary, you will need Jim Breen's Kanji Dictionary.
 *
 * This is free software published under the terms of the GNU public
 * license. copyright Adrian Daerr, Paris 2004
 */

import java.io.*;
import java.util.*;

public class CompJapanEngDict {
    Hashtable kdic;
    Vector nkanav = new Vector();
    final Character.UnicodeBlock HIRAGANA = Character.UnicodeBlock.HIRAGANA;
    final Character.UnicodeBlock KATAKANA = Character.UnicodeBlock.KATAKANA;
    BufferedWriter out;
    {
	System.err.println("Force UTF-8 encoding on stdout...");
	try {// bugware-alarm! The default encoding should be used
	    // instead of forcing utf-8; but how can the user change
	    // the default stdout encoding when launching the JavaVM ?
	    // LC_CTYPE appears to be ignored on MacOS 10.3.+Java 1.4.1
	    out = new BufferedWriter(new OutputStreamWriter(System.out,"UTF-8"));
	} catch (UnsupportedEncodingException e) {
	    System.err.println("Can't use explicit UTF-8 output, using stdout in default encoding");
	    out = new BufferedWriter(new OutputStreamWriter(System.out));
	}
    }

    public static void main(String args[])
        throws FileNotFoundException, IOException {
        if (args.length==0) {
            System.err.println("usage: CompJapanEngDict dict-file");
        } else {
	    new CompJapanEngDict().run(args);
        }
    }

    public void run(String args[])
        throws FileNotFoundException, IOException {
	int i;//the usual insignificant variable

        // open kanji dictionary
        readKanjiDict();

        // open japanese-english dictionary
        LineNumberReader jpendictf = new LineNumberReader(new InputStreamReader(new FileInputStream(args[0]),"UTF-8"));
        BufferedWriter outdictf = new BufferedWriter(new OutputStreamWriter(new FileOutputStream("out.txt"),"UTF-8"));

        // process line by line
	String ligne;
        while((ligne = jpendictf.readLine()) != null) {
            int posBeg,posEnd;
            // skip lines marked as comments
            if (ligne.startsWith("#") || ligne.length()==0) continue;
            posBeg = ligne.indexOf(" [");// look for KANA part
            if (posBeg < 0) {// conclude that KANJI is already pure KANA
                outdictf.write(ligne+"\n");
            } else {// calculate kanji ending markers
		// isolate and output kanji
                String kanji = ligne.substring(0,posBeg);
                outdictf.write(kanji);
                // substitute kanji repetition mark
		while ((i=kanji.indexOf("\u3005"))>0) {
		    kanji = kanji.substring(0,i)+kanji.charAt(i-1)+kanji.substring(i+1);
		}
		// count kanji
		int nKanji=0;
		for (i=0; i<kanji.length(); i++) {
		    if (! isKana(kanji.charAt(i))) nKanji++;
		}
                do {
                    posEnd = ligne.indexOf("]",posBeg);
                    //System.out.println("pos:"+posBeg+":"+posEnd);
                    String kana = ligne.substring(posBeg+2,posEnd);
		    if ((kana.indexOf(".")<0)&&(nKanji>1)) {
			nkanav.clear();//no separation marks yet
			findKanaSep(kanji,kana,"");
			if (nkanav.size()==0) {
			    // no decomposition found: be verbose...
			    println("line #"+(jpendictf.getLineNumber()-1)+
				    ": don't know how to decompose "+
				    kanji+" = "+kana);
			    for (i=0; i<kanji.length(); i++) {
				Object nK = (Object)new Character(kanji.charAt(i));
				if (kdic.containsKey(nK)) {
				    HashSet readings = (HashSet)((Vector)kdic.get(nK)).elementAt(1);
				    println("  with "+nK+" = "+readings.toString());
				}

			    }
			    // ... and write kana unchanged
			    outdictf.write(" ["+kana+"]");
			} else {
			    if (nkanav.size()>1) {
				// too many decompositions found
				println("line #"+(jpendictf.getLineNumber()-1)+
					": decomposition of "+kanji+
					" ambiguous (leaving undecided):");
				for (i=0; i<nkanav.size(); i++)
				    println("  (#"+i+") "+(String)nkanav.elementAt(i)+" ?");
				// write kana unchanged
				outdictf.write(" ["+kana+"]");
			    } else {
				// everything went smoothly
				outdictf.write(" ["+(String)nkanav.firstElement()+"]");
			    }
			}
		    } else {
			// nothing to be done
			if (nKanji>1) {
			    println("line #"+(jpendictf.getLineNumber()-1)+
				    ": already separated, copying unchecked: "+
				    kanji+" = "+kana);
			} else {
			    println("line #"+(jpendictf.getLineNumber()-1)+
				    ": trivial to decompose, copying: "+
				    kanji+" = "+kana);
			}
			outdictf.write(" ["+kana+"]");
		    }
                } while((posBeg = ligne.indexOf(" [",posBeg+1))>0);
                outdictf.write(ligne.substring(posEnd+1));
		// now calculate META-information
		int f=0, g=0;
		for (i=0; i<kanji.length(); i++) {
		    Object nK = (Object)new Character(kanji.charAt(i));
		    if (kdic.containsKey(nK)) {
			StringTokenizer meta = new StringTokenizer((String)((Vector)kdic.get(nK)).elementAt(3));
			while (meta.hasMoreTokens()) {
			    String st = meta.nextToken();
			    if (st.charAt(0)=='F') {//frequency-of-use ranking
				try {
				    f=Math.max(f,Integer.parseInt(st.substring(1),10));
				} catch (NumberFormatException e) {
				    System.err.println("line #"+(jpendictf.getLineNumber()-1)+
						       " NumberFormatException for F-field");
				    f=0;
				}
			    } else if (st.charAt(0)=='G') {//Jouyou-grade level
				try {
				    g=Math.max(g,Integer.parseInt(st.substring(1),10));
				} catch (NumberFormatException e) {
				    System.err.println("line #"+(jpendictf.getLineNumber()-1)+
						       " NumberFormatException for G-field");
				    g=0;
				}
			    }
			}
		    }

		}
		if (f>0) { outdictf.write(" <F"+f+">"); }
		if (g>0) { outdictf.write(" <G"+g+">"); }
                outdictf.write("\n");
            }

        }
        jpendictf.close();
        outdictf.close();

        //Enumeration e = pointsGPS.elements();
        //while (e.hasMoreElements())
        //    System.out.println(e.nextElement());
	out.close();
    }

    void readKanjiDict() 
        throws FileNotFoundException, IOException {
        kdic = new Hashtable(10000);
        // open kanji dictionary
        BufferedReader kdicf = new BufferedReader(new InputStreamReader(new FileInputStream("kanjidic.txt"),"UTF-8"));
        String ligne;
	Vector entry;// contains verbatim line and list of readings for kanji
        HashSet readings;// list of readings for given kanji
	StringBuffer meta;// information about kanji
	StringBuffer english;// english meanings

        // read kanji and definition
        while((ligne = kdicf.readLine()) != null) {
            // skip lines marked as comments
            if (ligne.startsWith("#") || ligne.length()<3) continue;
	    // find readings
	    readings = new HashSet();
	    meta = new StringBuffer();
	    english = new StringBuffer();
	    StringTokenizer st = new StringTokenizer(ligne);
	    st.nextToken();//skip kanji
	    while (st.hasMoreTokens()) {
		String token = st.nextToken();
		if (token.charAt(0)=='{') {// english meaning ?
		    english.append(" "+token);
		    while ((token.indexOf("}")<0) && (st.hasMoreTokens())) {
			token = st.nextToken();
			english.append(" "+token);
		    }
		    continue;
		}
		if (isKana(token.charAt(0))) {// yet another reading?
		    // convert to hiragana and add
		    StringBuffer hiraganaToken = new StringBuffer();
		    for (int i=0; i<token.length(); i++) {
			char c=token.charAt(i);
			if (isHiragana(c)) {
			    hiraganaToken.append(c);
			} else if (isKatakana(c)) {//convert to hiragana
			    hiraganaToken.append((char)(c-0x60));

			} else break;
		    }
		    token = hiraganaToken.toString();
		    readings.add(token);
		    // add possible modifications of readings
		    char c=token.charAt(0);
		    // tell if kana may soften (vocalize) in compounds
		    if (c=='\u304b' || //ka
			c=='\u304d' || //ki
			c=='\u304f' || //ku
			c=='\u3051' || //ke
			c=='\u3053' || //ko
			c=='\u3055' || //sa
			c=='\u3057' || //shi
			c=='\u3059' || //su
			c=='\u305b' || //se
			c=='\u305d' || //so
			c=='\u305f' || //ta
			c=='\u3061' || //chi
			c=='\u3064' || //tsu
			c=='\u3066' || //te
			c=='\u3068' || //to
			c=='\u306f' || //ha
			c=='\u3072' || //hi
			c=='\u3075' || //fu
			c=='\u3078' || //he
			c=='\u307b') {; //ho
			readings.add(""+((char)(token.charAt(0)+1))+
				     token.substring(1));
		    }
		    // h-series kana may harden in compounds (1)
		    if (c=='\u306f' || //ha
			c=='\u3072' || //hi
			c=='\u3075' || //fu
			c=='\u3078' || //he
			c=='\u307b') {; //ho
			readings.add(""+((char)(token.charAt(0)+2))+
				     token.substring(1));
		    }
		    // h-series kana may harden in compounds (2)
		    if (c=='\u3070' || //ba
			c=='\u3073' || //bi
			c=='\u3076' || //bu
			c=='\u3079' || //be
			c=='\u307c') {; //bo
			readings.add(""+((char)(token.charAt(0)+1))+
				     token.substring(1));
		    }
		    // trailing tsu tends to "become small"
		    if (token.charAt(token.length()-1)=='\u3064') {//tsu
			readings.add(token.substring(0,token.length()-1)+
				     '\u3063');
		    }
		} else {// meta-info
		    if (meta.length()>0) { meta.append(" "); }
		    meta.append(token);
		}
	    }
	    entry = new Vector(4);
	    entry.add((Object)ligne);
	    entry.add((Object)readings);
	    entry.add((Object)english.toString());
	    entry.add((Object)meta.toString());
	    //out.write(entry.toString()+"\n");
            kdic.put((Object)new Character(ligne.charAt(0)),(Object)entry);
        }
        kdicf.close();
        System.err.println("Read "+kdic.size()+" kanji entries");
	return;
    }

    void findKanaSep(String kanji, String kana, String sepkana)
	throws IOException {
	//println("kanji="+kanji+", kana="+kana+", sep="+sepkana);
	if ((kanji.length() == 0) || (kana.length() == 0)) {
	    if ((kanji.length() == 0) && (kana.length() == 0)) {
		nkanav.add(sepkana.endsWith(".")?
			   sepkana.substring(0,sepkana.length()-1):
			   sepkana);
	    }
	    return;
	}
	char nextKanjiChar = kanji.charAt(0);
	Character nextKanjiC = new Character(nextKanjiChar);
	String nextKanjiS = nextKanjiC.toString();
	if (isKana(nextKanjiChar)) {
	    if (kanji.charAt(0) == kana.charAt(0)) {
		findKanaSep(kanji.substring(1), kana.substring(1),
			    sepkana+nextKanjiS);
	    }
	} else if (kdic.containsKey((Object)nextKanjiC)) {
	    HashSet readings = (HashSet)((Vector)kdic.get((Object)nextKanjiC)).elementAt(1);
	    Iterator rIter = readings.iterator();
	    //println(readings.toString());
	    while (rIter.hasNext()) {
		String r = (String)rIter.next();
		if (kana.startsWith(r)) {
		    findKanaSep(kanji.substring(1), kana.substring(r.length()),
				sepkana+r+".");
		}
	    }
	} else {
	    println("character <"+nextKanjiS+"> neither a Kana nor in Kanji-Dict");
	    return;
	}
    }

    /** find out if character is HIRAGANA or KATAKANA */
    boolean isKana(char c) {
	final Character.UnicodeBlock type = Character.UnicodeBlock.of(c);
	return (type.equals(HIRAGANA) || type.equals(KATAKANA));
    }

    /** find out if character is HIRAGANA */
    boolean isHiragana(char c) {
	return Character.UnicodeBlock.of(c).equals(HIRAGANA);
    }

    /** find out if character is KATAKANA */
    boolean isKatakana(char c) {
	return Character.UnicodeBlock.of(c).equals(KATAKANA);
    }

    void println(String s) {
	try {
	    out.write(s);
	    out.newLine();
	} catch (IOException e) {
	    System.err.println("caught IOException while trying to print:");
	    System.err.println(s);
	    e.printStackTrace();
	}
    }
}


xletters.h
/* X Letters - Catch falling words */
/* Copyright (C) 1998 by
 *   Peter Horvai (peter.horvai@ens.fr)
 *   David A. Madore (david.madore@ens.fr)
 * Copyright (C) 2003 by
 *   Adrian Daerr (adrian.daerr@pmmh.espci.fr)
 */

/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See
 * the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */


/* The two following constants determine the size of the game space,
 * in pixels.  Note that they are intentionally not modifiable through
 * resources, as that would be cheating. */

/* The width of the play area, in pixels */
#define WIDTH 1792

/* The height of the play area, in pixels */
#define HEIGHT 1792



/* The following four constants are game parameters.  They do not
 * change the ``look'' of the game but rather its ``feel''. */

/* The maximal number of words simultaneously falling */
#define MAXWORDS 256

/* Whether the program should avoid at all cost the collision between
 * two words */
#define NOCOLLIDE 1

/* Whether the program should avoid at all cost the vertical collision
 * between two words (in case the deathmatch sends many words very
 * fast, or during a pause).  This is recommended. */
#define NOVCOLLIDE 1

/* Whether making a mistake will take you back to the last place in
 * the word that matches what you have just typed (rather than right
 * back to the beginning) */
/* broken since 2.0.0 ! */
#define CONTEXT 0

/* The duration of a half-level in time units. orig:600 */
#define LEVELDUR 600



/* The following constants are configuration variables */

/* The file from which non-bonus words are read */
#define WORDFILE "/tmp/xletters-words.utf8"

/* The file from which input-conversions are read */
#define ICONVFILE "/tmp/inputconv.txt"
/*#define ICONVFILE "/usr/local/share/xletters/inputconv.txt"*/

/* The high score file */
/* if you change this, you should probably also modify Makefile.am */
#define SCOREFILE "/var/local/games/lib/xletters/scores_jp-en"

/* Number of high score entries */
#define SCORELEN 20

/* Whether a single user is allowed to have several entries in the
 * high score table */
#define MULTENT 1


/* The following constants are default values for game resources */

/* Show kana next to word ? Sum desired behaviours to get parameter value:
 *  1 show typed part of furigana, in the lower part of the screen
 *  2 show typed part of furigana, always
 *  4 show untyped part of furigana, in the lower part of the screen
 *  8 show untyped part of furigana, always
 * 16 pick one furigana transcription at the beginning (i.e. 
 *    to destroy word, only this transcription will work). This implies +8 
 *    to avoid frustration. */
#define SHOWKANA 6

/* "lower part of the screen" in the above means y>SHOWKANA_Y */
#define SHOWKANA_Y (HEIGHT/2)

/* Show comment/translation next to word ? Possible values are:
 *  0 don't show comment/translation
 *  1 show comment/translation, in the lower part of the screen
 *  2 show comment/translation, always */
#define SHOWCOMMENT 2

/* The font to use, if not overriden by a resource specification */
#define FONTNAME "-*-*-*-*-*-ja-18-*-*-*-*-*-iso10646-*"

/* The color of the words, if not overriden by a resource
 * specification */
#define WORD_COLOR "Blue"

/* The color of the letters correctly typed, if not overriden by a
 * resource specification */
#define TYPED_COLOR "Red"

/* Same as previous, for the kana transcription */
#define FONTNAME "-*-*-*-*-*-ja-18-*-*-*-*-*-iso10646-*"
#define KANA_COLOR "CornflowerBlue"
#define KANATYPED_COLOR "Salmon"

/* And one more set of prefs for the translation */
#define COMMENTFONTNAME "-*-fixed-medium-r-normal-*-24-*-*-*-*-*-iso8859-1"
#define COMMENT_COLOR "LimeGreen"

#define KANJI_FREQ 0 /* limit to the ... most frequent kanji */

/* It is safe to leave the following untouched */

/* The base time unit, in milliseconds - you probably shoudln't change
 * this. */
#define BASETIME 75

/* The application class - you probably shouldn't change this. */
#define APP_CLASS "XLetters"
